### ISSUE TITLE
Enable time series graph event handler

### ISSUE THREAD

**Issue by smalers** [OpenWaterFoundation/owf-app-info-mapper-ng#82]
A common user interaction will be to click on a map feature and display a graph.  Other issues focus on features to display popup for mouse over, click on a feature, and link to external URLS.  However, this issue focuses on creating graphs within the Angular application.  Doing so involves implementing logic similar to the following:

1. Understand what is to be graphed.  The TSTool TSID concept is useful because it encapsulates the location ID, data type, interval, etc.  The TSTool "time series product" concept is also useful because it describes the appearance of a graph.  I previously enhanced TSTool to write the time series product file as JSON, envisioning that TSTool could be used to create graph templates for use with web applications.  These concepts can be implemented in the Info Mapper.
2. Retrieve/read the data for graphing.  This can occur by calling a web service, retrieving a file from a URL, or reading a local server file.  The initial approach will be to read a file from the local server, which will have been pre-generated, such as by TSTool.   For example, files containing only one time series may be stored in a server folder.  The format of the files will be CSV, JSON, or DateValue, as written by TSTool.  CSV is simple but lacks metadata.  DateValue is comprehensive and includes metadata but would require porting [Java Code](https://github.com/OpenCDSS/cdss-lib-common-java/blob/master/src/RTi/TS/DateValueTS.java) to TypeScript/JavaScript, JSON may be compromise in the short term and can be enhanced.
3. Graph the time series.  The Chart.js graph approach may be OK in the short term and if there are limitations, other packages can be used as per [owf-lib-viz-timeseries](https://github.com/OpenWaterFoundation/owf-lib-viz-timeseries) research.

To recap:

1. mouse over a feature to find the feature of interest
2. click on a feature to show a popup containing a link to display a graph
3. click on the link to display the graph
    1. read the associated graph definition from a template
    2. get the data for the graph
    3. hand off the data and template to code that creates the graph
    4. Graph displays, user interacts with and then closes

A lot of code has been previously written and can be leveraged.  Much of the work to implement involves figuring out how to describe functionality in configurations and integrating the existing code.

**Comment by Nightsphere** [comment]
I have pushed the new changes for this recently. This is how I have it working right now:

After cloning, npm install will install the chart.js zoom plugin.

The map-configuration file event handler has the following properties for clicking:
`"eventType": "click",`
`"action": "popup",`
`"templatePath": "../map-template-files/diversion-click.html",`
`"dataPath": "../../data-ts/graph-template-data.csv",`
`"configPath": "../../data-ts/graph-template.json"`

The path properties are all relative to the map configuration file right now. In order to also have
mouseovers, another mouseover event needs to be added, but the template is not used yet. As of
right now it just loops through all properties and shows those. An example map configuration file
that uses this is the [point-geometry-maps](../blob/master/info-mapper/src/assets/app-default/data-maps/map-configuration-files/point-geometry-maps.json).

The [popup template file](../blob/master/info-mapper/src/assets/app-default/data-maps/map-template-files/diversion-click.html) shows what properties to show on the initial popup from
Leaflet. As of right now, the only thing that can't change is the button html tag. The only way I
could get the button to work with Leaflet was assigning an id, so the id="internal-graph" and the
call to showGraph( ) have to be on the button.

The [data path file](../blob/master/info-mapper/src/assets/app-default/data-ts/graph-template-data.csv) is just the CSV file with the data for the graph.

The [config path file](../blob/master/info-mapper/src/assets/app-default/data-ts/graph-template.json) is the graph template JSON file, which needs the ${property} notation. For right
now, it needs to be `${featureProperties.property}`, where property is the name of the desired
property in that clicked feature.

The only three graph template properties I have working at the moment are:
  1. `product.properties.MainTitleString`
  2. `product.subProduct[0].properties.GraphType`
  3. `product.subProduct[0].properties.LeftYAxisTitleString`

So changing or updating these will update on the graph. I will work on more integration tomorrow.

**Comment by Nightsphere** [comment]
The data path is no longer in the map configuration file, but in the graph template file as we
discussed.

**Comment by smalers** [comment]
What Josh has done is a start but needs to be more sophisticated.  I'm thinking that we need to use JSON to define the template and its actions because we need more control, and asking people to write html/JSON templates will push their limits, especially for complex handling.  The InfoMapper can put together the HMTL/JS on the fly based on the configuration.  Some basic requirements are:

1. The popup might show the attributes from the map layer (similar to mouse-over popup), with control over which attributes and their formatting.
2. The popup might list one or more specific actions, such as displaying graphs.  Because more than one graph option may be offered, there needs to be a way to list those as buttons, choice, etc.  Buttons may be OK if the number of actions is small, but choice or other control may be needed if many actions are possible.

Ideally the JSON file could be created with GeoProcessor, TSTool, or other program.  The JSON file would have a format similar to the following (this is draft, not clean JSON).  Maybe the property should be 'popupTemplatePath' and use filename like 'diversion-popup-template.json'.

```
{
    "id" : "idOfPopupTemplate",  // Used internally
   "name": "Name of Popup Template",  // User may see this
   "description":  "List main attributes and provide buttons to graph time series.", 
    "layerAttributes" : {    // Control which attributes are listed and how they are formatted
          "include:" : [
                "*"
          ]
          "exclude" : [
         ]
         "formats": [  // Only include in the list of other than default formatting is needed
              {
                  "name" : "someAttriuteName",
                  "digits" :  number of digits for floating point.
                  "urlLabel": "what user sees".
                  "dateTimeFormat": "some pattern"
              }
         ]
    }
    "actions": [
         {
              "action" : "graph",   // causes graph code to run, but may come up with other actions
              "group":  "Graphs",   // Need some way to use list instead of buttons?
              "label" : "Graph abc",  // Results in button or appropriate UI component
              "productPath" : "what is currently called configPath",  // the template for the graph product
         }     
    ]
}
```

**Comment by Nightsphere** [comment]
I have pushed the changes for this. Here is the diversion-popup-template.json file I used:

```
{
  "id" : "diversion-popup-template",
  "name": "Diversion Template",
  "description":  "List main attributes and provide buttons to graph time series.",
  "layerAttributes" : {
    "include" : [ "*" ], // Will always include every attribute right now
    "exclude" : [],      // Not yet implemented
    "formats": []        // Not yet implemented
  },
  "actions": [
      {
        "action" : "graph",
        "group":  "Graphs",  // Not yet implemented
        "label" : "Graph Diversions",
        "productPath" : "../../data-ts/diversion-graph-template.json"
      }
  ]
}
```
Link detection should still be working as well.

**Comment by smalers** [comment]
I recommend the following changes:

* The `//` comments are not supported in JSON, right, so the above is only for illustration, right?  I will add to the documentation specification so we'll have documentation.
* Since the popup is not really a template anymore, change `popupTemplatePath` to `popupConfigPath` - does this make sense?  FIlenames will then be like `diversion-popup-config.json`.  The file could still be expanded to handle `${Property}` if we need to.
* In the TSTool graph template file, change the notation `${featureProperties.wdid}` to `${featureAttribute:wdid}`.  The text before the colon indicate a "scope" and we can use a period in the attribute name if we need to and also add "modifiers" like `.format(YYYY-MM-DD)` for date/times and `.format(%.4f)` for floating point
* The `TSID` property in the graph template should follow the full TSID standard.  See the [TSTool](http://opencdss.state.co.us/tstool/latest/doc-user/datastore-ref/StateMod/StateMod/) documentation.  For example, I'd like to use the string below with tildes, which will clearly indicate the identifier, type of file, and path to the file:

```
"TSID" : "${featureAttribute:wdid}.StateMod.Demand.Month~StateMod~/data-ts/diversions/${featureAttribute:wdid}.StateMod.Demand.Month.H.stm}"
```

I am probably going to add an attribute to point files named `nodeId` or something like that since `wdid` does not apply to all model nodes.

**Comment by smalers** [comment]
I made an attempt to configure a graph using the existing syntax and have an error:

![image](https://user-images.githubusercontent.com/6448274/85116510-ac4dfb00-b1da-11ea-9746-41cd4166e9d7.png)

I can't tell what is wrong from the message.  Maybe the paths are not correct as I have configured them.  The map configuration references the `graphs` folder:

```
              "eventHandlers": [
                {
                  "eventType": "click",
                  "name": null,
                  "description": "",
                  "properties": {
                    "popupTemplatePath": "graphs/diversion-popup-config.json"
                  }
                }
              ]
```

The popup configuration file in `graphs` folder has:

```
{
  "id" : "diversion-popup-template",
  "name": "Diversion Template",
  "description":  "List main attributes and provide buttons to graph time series.",
  "layerAttributes" : {
    "include" : [ "*" ],
    "exclude" : [],
    "formats": []
  },
  "actions": [
      {
        "action" : "graph",
        "label" : "Graph Demand",
        "productPath" : "diversion-demand-graph-template.json"
      },
      {
        "action" : "graph",
        "label" : "Graph Diversions",
        "productPath" : "diversion-historical-graph-template.json"
      }
  ]
}
```
It seems like it is not reading the popup configuration.  The TSTool files also exist in the `graphs` folder.

**Comment by smalers** [comment]
The error seemed to be most related to trying to use a property that does not exist.  For example, the desired behavior is:

If a property such as `${somePropertyA}` is requested but that property is not defined, then the output should literally have `${somePropertyA}` to indicate that the property needs to be defined.  Don't throw an error, don't fill in null, etc.

**Comment by Nightsphere** [comment]
`popupConfigPath` has now replaced `popupTemplatePath`.

Property notation is now in the form `${featureAttribute:wdid}`

Error messages should be more verbose now, and the TSID can take either the old 

`Location...Interval~/path/to/StateMod.stm`
or
`Location...Interval~StateMod~/path/to/StateMod.stm`

**Comment by smalers** [comment]
I believe that this functionality has been implemented.  I am going to close this issue.   We can open new issues as necessary.
